hardware_pydantic/devices.py

```diff
@@ -64,26 +114,114 @@ def post__transfer(
+            from_obj.content[vial.position] = None
         elif isinstance(from_obj, Heater):
             from_obj.content = None
         else:
             raise TypeError
-        transferee.position = None
-        transferee.position_relative = None
+        vial.position = None
+        vial.position_relative = None
         # put it to
         if isinstance(to_obj, Rack):
             assert to_position in to_obj.content
-            to_obj.content[to_position] = transferee.identifier
+            to_obj.content[to_position] = vial.identifier
         elif isinstance(to_obj, Heater):
-            to_obj.content = transferee.identifier
-        transferee.position = to_position
-        transferee.position_relative = to_obj.identifier
+            to_obj.content = vial.identifier
+        vial.position = to_position
+        vial.position_relative = to_obj.identifier
+
+
+class PlateGripperArm(Device):
+    pass
+
+
+class SolidSVTool(Device):
+    """Solid sample vial tool for transferring solids."""
+    # todo: this device is often used together with a balance
+    # note: this is not used for the current simulation
+    def pre__transfer(self, from_obj: Vial, to_obj: Vial, amount: float):
+        """Transfer solids from one vial to another."""
+        # TODO sample from a dist
+        removed = from_obj.remove_content(amount)
+        to_obj.add_content(removed)
+
+    def post__transfer(self, from_obj: Vial, to_obj: Vial, amount: float):
+        """Time needed to transfer solids from one vial to another."""
+        # todo: add adaptive post to estimate the time of transferring solids
+        # transfer speed in g/s
+        transfer_speed = 2
+        return amount / transfer_speed
+
+
+class Balance(Device):
+    """Balance for weighing solids."""
+    reading: float = 0
+    content: LabObject | None = None
+
+    def pre__tare(self):
+        """Set the reading to zero."""
+        self.reading = 0
+
+    def post__tare(self):
+        """Time needed to tare the balance."""
+        return 1.e-6
+
+    def pre__weigh(self, amount: float):
+        """Set the reading to the amount."""
+        self.reading = amount
+
+    def post__weigh(self):
+        """Time needed to weigh solids."""
+        # todo: add adaptive post to estimate the time of weighing solids
+        # assuming the weighing takes 10 seconds
+        return 10
+
+
+class Evaporator:
+    """Evaporator for evaporating solvents."""
+    # denotes v-10 evaporator for the moment
+    reading: dict[str, float] = {
+        # temperature in Celsius
+        "temperature": 25,
+        # pressure in atm
+        "pressure": 1,
+        # rpm
+        "rpm": 0,
+    }
+    set_point: dict[str, float] = {
+        # temperature in Celsius
+        "temperature": 25,
+        # pressure in atm
+        "pressure": 1,
+        # rpm
+        "rpm": 0,
+    }
+    content: LabObject | None = None
+
+    def pre__set_point(self, set_point: dict[str, float]):
```

Comment:
It you look at my type hinting for other `pre__` access, this method should always return a list of will be occupied `LabObject` and a float representing projected time cost.